drivers/include/eeprom.h

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (C) 2018 Inria
+ *               2019 Otto-von-Guericke-Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @defgroup    drivers_eeprom Common EEPROM interface
+ * @ingroup     drivers_storage
+ *
+ * This module provides a common API for accessing EEPROM (both internal and
+ * external), as well as common code and convenience interfaces.
+ *
+ * @{
+ *
+ * @file
+ * @brief       Interface definition of EEPROM drivers
+ *
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ * @author      Marian Buschsieweke <marian.buschsieweke@ovgu.de>
+ */
+
+#ifndef EEPROM_H
+#define EEPROM_H
+
+#include <stddef.h>
+#include <stdint.h>
+#include <unistd.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef HAVE_EEPROM_T
+/**
+ * @brief   Device handle type for EEPROM devices
+ */
+typedef uint_fast8_t eeprom_t;
+#endif
+
+/**
+ * @brief   Return the size of the EEPROM device
+ *
+ * @param[in]       dev     Device handle of the EEPROM device to get the size
+ *
+ * @return  The size of the EEPROM device in bytes
+ * @retval  -1              No EEPROM device for parameter @p dev
+ */
+ssize_t eeprom_size(eeprom_t dev);
```

Comment:
On 16-bit platforms, the EEPROM size might actually be bigger that `ssize_t`

Reply:
I changed it to use `size_t`, so that up to 64 KiB can be used on 8-bit and 16-bit platforms. I'm not sure if it makes more sense to use `uint32_t` as the current EEPROM-API does, or the additional overhead on 8-bit and 16-bit platforms is not worth it.

Maybe introducing `eeprom_off_t` for size and positions is a solution, such as:

```C
#ifndef HAVE_EEPROM_OFF_T
typedef size_t eeprom_off_t;
#endif
```

Offloading decisions to the user is always a good idea :-D

Reply:
Well I have here a 1MBit SPI EEPROM, so `size_t` is still too small. In fact I only noticed this because CI would fail on those platforms (I used `size_t` too at first.)

I don't know how big the overhead of just always using `uint32_t` is though.

In my driver I have a configurable [`AT25XXX_PARAM_ADDR_LEN`](https://github.com/RIOT-OS/RIOT/pull/11945/files#diff-0db8ad96013abefa0eb5a2660166ec68R53), we could make it `EEPROM_ADDR_LEN` and set the type of `eeprom_off_t` based on that.
It will make the code a bit more difficult though.

Reply:
> I don't know how big the overhead of just always using `uint32_t` is though.

Likely less than the overhead of the indirect function call via the `eeprom_driver_t` (but I honestly don't know and would have to disassemble the code to see). I added the `eeprom_off_t` and it defaults to `uin32_t`. Less maintenance headache by using the `uint32_t` is likely worth the overhead. And having a `typedef` in place, this could still be optimized later on without changing the API.